src/v/cloud_topics/reconciler/reconciler.cc

```diff
@@ -0,0 +1,310 @@
+            // time to do some work
+        }
+
+        try {
+            co_await reconcile();
+        } catch (...) {
+            vlog(
+              lg.info,
+              "Recoverable error during reconciliation: {}",
+              std::current_exception());
+        }
+    }
+}
+
+ss::future<> reconciler::reconcile() {
+    auto object = co_await build_object();
+    if (!object.has_value()) {
+        co_return;
+    }
+
+    auto result = co_await upload_object(std::move(object->data));
+    if (result != cloud_io::upload_result::success) {
+        vlog(lg.info, "Failed to upload L1 object: {}", result);
+        co_return;
+    }
+
+    // commit for each partition represented in the uploaded object
+    for (const auto& range : object->ranges) {
+        co_await commit_object(range);
+    }
+}
+
+ss::future<std::optional<reconciler::object>> reconciler::build_object() {
+    // light-weight copy for stable iteration
+    std::vector<attached_partition> partitions;
+    for (const auto& p : _partitions) {
+        if (p.second->partition->is_leader()) {
+            partitions.push_back(p.second);
+        }
+    }
+
+    // avoid starving partitions
+    std::shuffle(
+      partitions.begin(), partitions.end(), random_generators::internal::gen);
+
+    object object;
+    auto size_budget = max_object_size;
+    for (const auto& partition : partitions) {
+        auto reader = co_await make_reader(partition, size_budget);
+        auto range = co_await std::move(reader).consume(
+          range_batch_consumer{}, model::no_timeout);
+        if (range.has_value()) {
+            object.add(std::move(*range), partition);
+            size_budget -= std::min(object.data.size_bytes(), size_budget);
+        }
+    }
+
+    if (object.data.empty()) {
+        co_return std::nullopt;
+    }
+
+    co_return object;
+}
+
+ss::future<cloud_io::upload_result> reconciler::upload_object(iobuf payload) {
+    const cloud_storage_clients::object_key key(
+      fmt::format("l1_{}", uuid_t::create()));
+
+    retry_chain_node rtc(
+      _as,
+      ss::lowres_clock::now() + std::chrono::seconds(2),
+      std::chrono::seconds(1));
+
+    co_return co_await _cloud_io->local().upload_object({
+      .transfer_details = {
+        .bucket = _bucket,
+        .key = key,
+        .parent_rtc = rtc,
+      },
+      .display_str = "l1_object",
+      .payload = std::move(payload),
+    });
+}
+
+ss::future<> reconciler::commit_object(const object_range_info& range) {
+    /*
+     * build and replicate a overlay batch to the underlying partition that
+     * records the range of data in the newly uploaded L1 object.
+     */
+    auto reader = [] {
+        /*
+         * TODO: as you can see we are replicating a version_fence batch. this
+         * is a temporary replacement for the real overlay batch type.
+         */
+        storage::record_batch_builder builder(
+          model::record_batch_type::version_fence, model::offset(0));
```

Comment:
as you can see we are replicating a version_fence batch. this is a temporary replacement for the real overlay batch type. it's convenient for testing because fence batch type is ignored by offset translation.